### ISSUE TITLE
Crash on Android when there is no app for file extension

### ISSUE THREAD

**Issue by Dancovich** [miguelpruivo/plugins_flutter_file_picker#97]
On Android if the device doesn't have an app that can respond to the generated Intent the plugin will cause the app to crash instead of returning the error to Dart or dealing with the exception itself.

I'm using version 1.3.6.

How to reproduce:

- Create an emulated device with API 16 and without Google Play Services (or use a device with no apps that can deal with application/pdf MIME type).
- On the Dart code call `await FilePicker.getFilePath(type: FileType.CUSTOM, fileExtension: "PDF");`
- App will crash with the stack trace bellow

```
I/FilePicker( 3752): Custom file type: application/pdf
E/MethodChannel#file_picker( 3752): Failed to handle method call
E/MethodChannel#file_picker( 3752): android.content.ActivityNotFoundException: No Activity found to handle Intent { act=android.intent.action.PICK cat=[android.intent.category.OPENABLE] typ=application/pdf (has extras) }
E/MethodChannel#file_picker( 3752): 	at android.app.Instrumentation.checkStartActivityResult(Instrumentation.java:1545)
E/MethodChannel#file_picker( 3752): 	at android.app.Instrumentation.execStartActivity(Instrumentation.java:1416)
E/MethodChannel#file_picker( 3752): 	at android.app.Activity.startActivityForResult(Activity.java:3351)
E/MethodChannel#file_picker( 3752): 	at android.app.Activity.startActivityForResult(Activity.java:3312)
E/MethodChannel#file_picker( 3752): 	at com.mr.flutter.plugin.filepicker.FilePickerPlugin.startFileExplorer(FilePickerPlugin.java:188)
E/MethodChannel#file_picker( 3752): 	at com.mr.flutter.plugin.filepicker.FilePickerPlugin.onMethodCall(FilePickerPlugin.java:126)
E/MethodChannel#file_picker( 3752): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:201)
E/MethodChannel#file_picker( 3752): 	at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(DartMessenger.java:88)
E/MethodChannel#file_picker( 3752): 	at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(FlutterJNI.java:219)
E/MethodChannel#file_picker( 3752): 	at android.os.MessageQueue.nativePollOnce(Native Method)
E/MethodChannel#file_picker( 3752): 	at android.os.MessageQueue.next(MessageQueue.java:125)
E/MethodChannel#file_picker( 3752): 	at android.os.Looper.loop(Looper.java:124)
E/MethodChannel#file_picker( 3752): 	at android.app.ActivityThread.main(ActivityThread.java:4745)
E/MethodChannel#file_picker( 3752): 	at java.lang.reflect.Method.invokeNative(Native Method)
E/MethodChannel#file_picker( 3752): 	at java.lang.reflect.Method.invoke(Method.java:511)
E/MethodChannel#file_picker( 3752): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786)
E/MethodChannel#file_picker( 3752): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553)
E/MethodChannel#file_picker( 3752): 	at dalvik.system.NativeStart.main(Native Method)
```

**Comment by miguelpruivo** [comment]
Hi @Dancovich. Well, I believe this is the desired behavior. Prior to `1.3.6` I was handling the exceptions in the plugin itself without rethrowing it to the user app's layer, now you can handle it yourself despite being interpreted in the plugin before.

If the device can't handle that file type, is expected to throw an exception to let you know, and handle it accordingly. You need to add a `try-catch` to the plugin call and deal with the exception yourself.

This a though implementation decision because there are users that want exceptions to be thrown so they can deal it them, and there are others that would rather want to not have to worry with it.

**Comment by Dancovich** [comment]
I did several tests, tried running `flutter clean`, made absolutely sure I was using version 1.3.6 and even took a look into your code (indeed you do capture the exception and just return a specific message I'm assuming you treat on your dart code and rethrows the exception).

Unfortunately the code still crashes on the specified platform even though I'm placing my code inside a try/catch block. The system shows a dialog with the message "Unfortunately, App has stopped."

Here's the exact code that's causing the crash:
```dart
try {
      String path = await FilePicker.getFilePath(
          type: FileType.CUSTOM, fileExtension: "PDF");
      if (path?.isNotEmpty != true) {
        return;
      }
    } catch (e) {
      print("it crashed");
      return;
    }
```
The `print("it crashed")` call does run but the app crashes nonetheless, giving me no opportunity to treat the error on Flutter.

If this isn't happening in your end maybe that's an issue specific to API 16 or specific to an emulator running API 16. I'll try creating emulator instances with other API levels to see if that changes the outcome.

**Comment by miguelpruivo** [comment]
Did you try to also catch the `PlatformException` on your side?
```
try{
 } on PlatformException catch (e) {
   // Do something...
   return;
```

**Comment by Dancovich** [comment]
About the version, I was running 1.3.5 before and it crashed the same way. I actually updated to 1.3.6 in hopes it would fix the issue but alas (although I'm not entirely convinced that the SDK isn't just trolling be and failing to do a clean build).

About the exception, I tried with your code and got the same result: app crashed.

I'm still new to Flutter, is there a difference between specifying the exception and just using `catch(e)`? I was under the impression `catch(e)` would catch every single exception.

Here's a complete stacktrace after your change. One thing I noticed is that the app crashes before printing "it crashed" (the dialog appears before returning to my app). In fact if I place a breakpoint inside the catch block it will stop there while the dialog is shown but as soon as I continue running or close the dialog the debugger will lose connection to the app.

```
I/FilePicker( 4256): Custom file type: application/pdf
E/MethodChannel#file_picker( 4256): Failed to handle method call
E/MethodChannel#file_picker( 4256): android.content.ActivityNotFoundException: No Activity found to handle Intent { act=android.intent.action.PICK cat=[android.intent.category.OPENABLE] typ=application/pdf (has extras) }
E/MethodChannel#file_picker( 4256): 	at android.app.Instrumentation.checkStartActivityResult(Instrumentation.java:1545)
E/MethodChannel#file_picker( 4256): 	at android.app.Instrumentation.execStartActivity(Instrumentation.java:1416)
E/MethodChannel#file_picker( 4256): 	at android.app.Activity.startActivityForResult(Activity.java:3351)
E/MethodChannel#file_picker( 4256): 	at android.app.Activity.startActivityForResult(Activity.java:3312)
E/MethodChannel#file_picker( 4256): 	at com.mr.flutter.plugin.filepicker.FilePickerPlugin.startFileExplorer(FilePickerPlugin.java:188)
E/MethodChannel#file_picker( 4256): 	at com.mr.flutter.plugin.filepicker.FilePickerPlugin.onMethodCall(FilePickerPlugin.java:126)
E/MethodChannel#file_picker( 4256): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:201)
E/MethodChannel#file_picker( 4256): 	at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(DartMessenger.java:88)
E/MethodChannel#file_picker( 4256): 	at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(FlutterJNI.java:219)
E/MethodChannel#file_picker( 4256): 	at android.os.MessageQueue.nativePollOnce(Native Method)
E/MethodChannel#file_picker( 4256): 	at android.os.MessageQueue.next(MessageQueue.java:125)
E/MethodChannel#file_picker( 4256): 	at android.os.Looper.loop(Looper.java:124)
E/MethodChannel#file_picker( 4256): 	at android.app.ActivityThread.main(ActivityThread.java:4745)
E/MethodChannel#file_picker( 4256): 	at java.lang.reflect.Method.invokeNative(Native Method)
E/MethodChannel#file_picker( 4256): 	at java.lang.reflect.Method.invoke(Method.java:511)
E/MethodChannel#file_picker( 4256): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786)
E/MethodChannel#file_picker( 4256): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553)
E/MethodChannel#file_picker( 4256): 	at dalvik.system.NativeStart.main(Native Method)
D/AndroidRuntime( 4256): Shutting down VM
W/dalvikvm( 4256): threadid=1: thread exiting with uncaught exception (group=0xa66a1228)
E/AndroidRuntime( 4256): FATAL EXCEPTION: main
E/AndroidRuntime( 4256): java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=13147, result=0, data=null} to activity {br.com.cosmiceffect.app.cosmiccollection.dev/br.com.cosmiceffect.app.cosmiccollection.MainActivity}: java.lang.IllegalStateException: Reply already submitted
E/AndroidRuntime( 4256): 	at android.app.ActivityThread.deliverResults(ActivityThread.java:3141)
E/AndroidRuntime( 4256): 	at android.app.ActivityThread.handleSendResult(ActivityThread.java:3184)
E/AndroidRuntime( 4256): 	at android.app.ActivityThread.access$1100(ActivityThread.java:130)
E/AndroidRuntime( 4256): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1243)
E/AndroidRuntime( 4256): 	at android.os.Handler.dispatchMessage(Handler.java:99)
E/AndroidRuntime( 4256): 	at android.os.Looper.loop(Looper.java:137)
E/AndroidRuntime( 4256): 	at android.app.ActivityThread.main(ActivityThread.java:4745)
E/AndroidRuntime( 4256): 	at java.lang.reflect.Method.invokeNative(Native Method)
E/AndroidRuntime( 4256): 	at java.lang.reflect.Method.invoke(Method.java:511)
E/AndroidRuntime( 4256): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786)
E/AndroidRuntime( 4256): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553)
E/AndroidRuntime( 4256): 	at dalvik.system.NativeStart.main(Native Method)
E/AndroidRuntime( 4256): Caused by: java.lang.IllegalStateException: Reply already submitted
E/AndroidRuntime( 4256): 	at io.flutter.embedding.engine.dart.DartMessenger$Reply.reply(DartMessenger.java:124)
E/AndroidRuntime( 4256): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler$1.success(MethodChannel.java:204)
E/AndroidRuntime( 4256): 	at com.mr.flutter.plugin.filepicker.FilePickerPlugin$1.onActivityResult(FilePickerPlugin.java:93)
E/AndroidRuntime( 4256): 	at io.flutter.app.FlutterPluginRegistry.onActivityResult(FlutterPluginRegistry.java:222)
E/AndroidRuntime( 4256): 	at io.flutter.app.FlutterActivityDelegate.onActivityResult(FlutterActivityDelegate.java:142)
E/AndroidRuntime( 4256): 	at io.flutter.app.FlutterActivity.onActivityResult(FlutterActivity.java:142)
E/AndroidRuntime( 4256): 	at android.app.Activity.dispatchActivityResult(Activity.java:5192)
E/AndroidRuntime( 4256): 	at android.app.ActivityThread.deliverResults(ActivityThread.java:3137)
E/AndroidRuntime( 4256): 	... 11 more
I/flutter ( 4256): [FilePicker] Platform exception: PlatformException(error, No Activity found to handle Intent { act=android.intent.action.PICK cat=[android.intent.category.OPENABLE] typ=application/pdf (has extras) }, null)
I/flutter ( 4256): it crashed
```

**Comment by miguelpruivo** [comment]
@Dancovich yes, there's actually a difference. PlatformException is made to communicate a fail with a platform plugin (see more [here](https://api.flutter.dev/flutter/services/PlatformException-class.html)) however, it's happening because the underlying (Android) platform is killing the app, that's why you can't prevent it.

I'll look into it, and fix it in the next version (aka `1.3.7`). 

Thank you for reporting.

**Comment by Dancovich** [comment]
Thanks, if you need more information feel free to message me.

**Comment by miguelpruivo** [comment]
Just released a fix for this in `beta`. Could you please give it a try and let me know if it's working for you? Just replace the `file_picker: 1.3.6` in your pubspec.yaml with the following:
```
  file_picker:
    git:
      url: https://github.com/miguelpruivo/plugins_flutter_file_picker.git
      ref: beta
```

Thank you!

**Comment by Dancovich** [comment]
@miguelpruivo Worked like a charm! Now catching a PlatformException will handle the exception with code `FilePicker` and the message `"Can't handle the provided file type."`.

**Comment by miguelpruivo** [comment]
@Dancovich glad to know. I'll deploy it on pub.dev later this week in an upcoming version, I'm just working in another issues as well.

Just let me know if you experience any other issue.

**Comment by miguelpruivo** [comment]
@Dancovich glad to know. I'll deploy it on pub.dev later this week in an upcoming version, I'm just working in another issues as well.

Just let me know if you experience any other issue.
Thank you.

**Comment by miguelpruivo** [comment]
Fixed in [`1.3.7`](https://pub.dev/packages/file_picker#137).
